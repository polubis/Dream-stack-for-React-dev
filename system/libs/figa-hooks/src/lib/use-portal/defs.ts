import type { ReactNode, ReactPortal } from 'react';

type RenderPortal = (children: ReactNode) => ReactPortal | null;

type UsePortal = () => {
  render: RenderPortal;
};

export type { UsePortal, RenderPortal };
