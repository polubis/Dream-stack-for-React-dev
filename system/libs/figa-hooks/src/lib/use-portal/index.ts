export * from './defs';
export * from './use-portal';
