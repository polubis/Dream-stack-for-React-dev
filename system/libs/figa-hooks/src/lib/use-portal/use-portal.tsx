import { useEffect, useMemo } from 'react';
import { createPortal } from 'react-dom';
import type { UsePortal } from './defs';

// usePortal is an implementation of the facade pattern.
const usePortal: UsePortal = () => {
  // Creates only one instance of div.
  const wrapper = useMemo(() => document.createElement('div'), []);

  useEffect(() => {
    // Adds div tag to body.
    document.body.appendChild(wrapper);

    return () => {
      // After unmounting the component - removes the div created earlier.
      document.body.removeChild(wrapper);
    };
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Returns an object with function that will allow you to use the portal.
  return {
    // This anonymous function is an implementation of the factory method pattern.
    render: (children) => createPortal(children, wrapper),
  };
};

export { usePortal };
