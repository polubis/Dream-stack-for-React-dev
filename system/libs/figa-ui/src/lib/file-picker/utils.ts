import type { ChangeEvent } from 'react';
import type { FilePickerPreviewList } from './defs';

const loadImage = (file: File): Promise<string> => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onloadend = () => {
      if (typeof reader.result !== 'string') {
        reject();
        return;
      }

      resolve(reader.result);
    };

    reader.readAsDataURL(file);
  });
};

const loadPreview = async (
  e: ChangeEvent<HTMLInputElement>
): Promise<[FileList | null, FilePickerPreviewList]> => {
  const { files } = e.target;

  if (!files) {
    return [null, []];
  }

  const promises: Promise<string>[] = [];

  for (let i = 0; i < files.length; i++) {
    const file = files[i];
    promises.push(loadImage(file));
  }

  const preview = await Promise.all(promises);

  return [files, preview];
};

export { loadPreview };
