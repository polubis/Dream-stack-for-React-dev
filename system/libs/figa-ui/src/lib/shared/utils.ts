import { type SpacingKey, tokens } from '../theme-provider';

const spacing = {
  get: (key: SpacingKey): string => tokens.spacing[key],
  parse: (key: SpacingKey): SpacingKey =>
    Number.parseInt(tokens.spacing[key].replace('px', '')) as SpacingKey,
};

export { spacing };
