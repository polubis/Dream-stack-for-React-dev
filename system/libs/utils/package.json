{
  "name": "@system/utils",
  "version": "0.0.1",
  "type": "commonjs"
}
