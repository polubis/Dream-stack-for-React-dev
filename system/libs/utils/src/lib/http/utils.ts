const isUrl = (url?: string): url is string => url !== undefined;

const getLackOfUrlError = (): string => {
  return `
    "Url" parameter is "undefined". API cannot work without url...

    It may be caused because: 
        - you used environment variable which is not specified,
        - you forget to pass url parameter and ignored TypeScript.
  `;
};

export { isUrl, getLackOfUrlError };
