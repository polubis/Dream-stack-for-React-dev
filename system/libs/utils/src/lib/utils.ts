const utils = (): string => {
  return 'utils';
};

export { utils };
