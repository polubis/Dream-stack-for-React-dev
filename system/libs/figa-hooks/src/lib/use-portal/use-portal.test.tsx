import { renderHook } from '@testing-library/react';
import { usePortal } from './use-portal';
import { isServer } from '@system/utils';

jest.mock('@system/utils');

describe('Portal is created when: ', () => {
  it('does nothing and returns null on the server side', () => {
    (isServer as jest.Mock).mockReturnValue(true);
    const { result } = renderHook(() => usePortal());

    expect(result.current.render(<div>Content</div>)).toBe(null);
  });

  it('moves passed JSX node outside of the main application wrapper', () => {
    (isServer as jest.Mock).mockReturnValue(false);
    const { result } = renderHook(() => usePortal());

    expect(result.current.render(<div>Content</div>)).toMatchSnapshot();
  });
});
