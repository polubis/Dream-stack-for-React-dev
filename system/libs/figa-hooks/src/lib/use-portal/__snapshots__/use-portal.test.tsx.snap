// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal is created when:  moves passed JSX node outside of the main application wrapper 1`] = `
Object {
  "$$typeof": Symbol(react.portal),
  "children": <div>
    Content
  </div>,
  "containerInfo": <div />,
  "implementation": null,
  "key": null,
}
`;
